File: src/engine/IDTradingEngine.js
import { spTime } from '../shared/utils.js';

/**
 * ID Trading Engine
 * 
 * Implements Intraday trading with gate closures:
 * - DA results lock in initial positions for all 48 SPs of tomorrow
 * - ID trading allows individual SP adjustments until gate closure
 * - Gate closures happen progressively:
 *   SP1: 23:00 D (Day before)
 *   SP2: 23:30 D
 *   ...
 *   SP20: 08:30 D+1
 *   SP48: 22:30 D+1
 * - BM_GATE locks 1 hour before each SP's delivery
 * - Players trade SP-by-SP with per-SP buy/sell buttons
 */

// ─── GATE CLOSURE TIMING ───
// Each SP has a gate closure time (in hours from start of Day D)
// SP1 (00:00-00:30 D+1) closes at 23:00 D
// SP2 (00:30-01:00 D+1) closes at 23:30 D
// ...
// SP48 (23:30-00:00 D+2) closes at 22:30 D+1

export function getGateClosureHour(sp) {
  // SP1-20: closes on Day D (23:00, 23:30, ... 08:30)
  // SP21-48: closes on Day D+1 (09:00, 09:30, ... 22:30)
  if (sp >= 1 && sp <= 20) {
    // 23:00 for SP1, 23:30 for SP2, etc.
    return 23 + (sp - 1) * 0.5;
  } else {
    // SP21 = 09:00, SP22 = 09:30, ... SP48 = 22:30
    return 9 + (sp - 21) * 0.5;
  }
}

export function getBMGateClosureHour(sp) {
  // BM gate closes 1 hour before SP delivery
  // SP1 delivery at 00:00, BM gate closes at 23:00 previous day
  const spStartHour = ((sp - 1) * 0.5) % 24;
  let bmGateHour = spStartHour - 1;
  if (bmGateHour < 0) {
    bmGateHour += 24; // Previous day
  }
  
  // Adjust for day boundary
  if (sp <= 2) {
    return 23 + (sp - 1) * 0.5; // Day D
  } else {
    return bmGateHour; // Day D+1
  }
}

export function formatGateClosureTime(sp) {
  const hour = getGateClosureHour(sp);
  const day = sp <= 20 ? "TODAY" : "TOMORROW";
  const h = Math.floor(hour) % 24;
  const m = (hour % 1) * 60;
  return `${String(h).padStart(2, '0')}:${String(Math.round(m)).padStart(2, '0')} ${day}`;
}

// ─── GATE STATUS ───

/**
 * Check if a specific SP is open for ID trading
 * @param {number} sp - Settlement period (1-48)
 * @param {number} currentTimeHour - Current game time in hours from Day D start
 * @returns {boolean}
 */
export function isIDGateOpen(sp, currentTimeHour) {
  const gateHour = getGateClosureHour(sp);
  // If SP is on Day D+1 (sp > 20), adjust currentTimeHour to account for day rollover
  if (sp > 20) {
    // Current time needs to be Day D+1 hours (24+) to compare
    return currentTimeHour >= 24 && currentTimeHour < gateHour + 24;
  }
  return currentTimeHour < gateHour;
}

/**
 * Get time remaining until gate closure
 * @param {number} sp - Settlement period (1-48)
 * @param {number} currentTimeHour - Current game time in hours
 * @returns {number} Hours remaining (negative if closed)
 */
export function getTimeToGateClosure(sp, currentTimeHour) {
  const gateHour = getGateClosureHour(sp);
  if (sp > 20) {
    return (gateHour + 24) - currentTimeHour;
  }
  return gateHour - currentTimeHour;
}

/**
 * Get all open SPs for ID trading
 * @param {number} currentTimeHour - Current game time in hours
 * @returns {Array} Array of open SP numbers
 */
export function getOpenSPs(currentTimeHour) {
  const open = [];
  for (let sp = 1; sp <= 48; sp++) {
    if (isIDGateOpen(sp, currentTimeHour)) {
      open.push(sp);
    }
  }
  return open;
}

/**
 * Get recently closed SPs (for notifications)
 * @param {number} currentTimeHour - Current game time
 * @param {number} prevTimeHour - Previous game time
 * @returns {Array} SPs that just closed
 */
export function getNewlyClosedSPs(currentTimeHour, prevTimeHour) {
  const newlyClosed = [];
  for (let sp = 1; sp <= 48; sp++) {
    const wasOpen = isIDGateOpen(sp, prevTimeHour);
    const isClosed = !isIDGateOpen(sp, currentTimeHour);
    if (wasOpen && isClosed) {
      newlyClosed.push(sp);
    }
  }
  return newlyClosed;
}

// ─── ID TRADE STRUCTURE ───

export function createIDTrade(sp, playerId, side, volumeMW, price, timestamp) {
  return {
    id: `id_${Date.now()}_${Math.random().toString(36).slice(2, 5)}`,
    sp,
    playerId,
    side, // 'buy' or 'sell'
    volumeMW: Math.abs(volumeMW),
    price,
    timestamp,
    status: 'pending' // pending, confirmed, rejected
  };
}

// ─── POSITION TRACKING ───

/**
 * Calculate current position for each SP after DA + ID trades
 * @param {Array} daVolumes - Array of 48 DA volumes (positive=buy/accept, negative=sell/generate)
 * @param {Array} idTrades - Array of ID trade objects
 * @param {string} playerId - Player ID
 * @returns {Array} Array of 48 position objects
 */
export function calculatePositions(daVolumes, idTrades, playerId) {
  const positions = [];
  
  for (let sp = 1; sp <= 48; sp++) {
    const daVol = daVolumes[sp - 1] || 0;
    const spTrades = idTrades.filter(t => t.sp === sp && t.playerId === playerId && t.status === 'confirmed');
    
    const idBuyVol = spTrades
      .filter(t => t.side === 'buy')
      .reduce((sum, t) => sum + t.volumeMW, 0);
    const idSellVol = spTrades
      .filter(t => t.side === 'sell')
      .reduce((sum, t) => sum + t.volumeMW, 0);
    
    const netPosition = daVol + idBuyVol - idSellVol;
    
    positions.push({
      sp,
      daVolume: daVol,
      idBuyVolume: idBuyVol,
      idSellVolume: idSellVol,
      netPosition,
      trades: spTrades,
      isOpen: true // Will be updated by caller with current time
    });
  }
  
  return positions;
}

// ─── ID MARKET CLEARING ───

/**
 * Simple continuous matching for ID trades
 * Matches buy and sell orders at overlapping prices
 * @param {Array} buyOrders - Buy orders [{ playerId, volumeMW, price, sp }]
 * @param {Array} sellOrders - Sell orders [{ playerId, volumeMW, price, sp }]
 * @returns {Array} Matched trades [{ buyOrder, sellOrder, matchVolume, matchPrice }]
 */
export function matchIDOrders(buyOrders, sellOrders) {
  const matches = [];
  
  // Sort buys high price first, sells low price first
  const sortedBuys = [...buyOrders].sort((a, b) => b.price - a.price);
  const sortedSells = [...sellOrders].sort((a, b) => a.price - b.price);
  
  for (const buy of sortedBuys) {
    for (const sell of sortedSells) {
      // Match if buy price >= sell price
      if (buy.price >= sell.price && buy.volumeMW > 0 && sell.volumeMW > 0) {
        const matchVolume = Math.min(buy.volumeMW, sell.volumeMW);
        const matchPrice = (buy.price + sell.price) / 2; // Midpoint
        
        matches.push({
          sp: buy.sp,
          buyerId: buy.playerId,
          sellerId: sell.playerId,
          volumeMW: matchVolume,
          price: matchPrice
        });
        
        buy.volumeMW -= matchVolume;
        sell.volumeMW -= matchVolume;
      }
    }
  }
  
  return matches;
}

// ─── UI HELPERS ───

export function getGateStatusDisplay(sp, currentTimeHour) {
  const isOpen = isIDGateOpen(sp, currentTimeHour);
  const hoursRemaining = getTimeToGateClosure(sp, currentTimeHour);
  
  if (!isOpen) {
    return {
      status: 'closed',
      message: 'LOCKED - No more ID trades',
      color: '#f0455a',
      canTrade: false
    };
  }
  
  if (hoursRemaining <= 0.5) {
    return {
      status: 'urgent',
      message: `${Math.round(hoursRemaining * 60)}m left - URGENT`,
      color: '#f5b222',
      canTrade: true
    };
  }
  
  if (hoursRemaining <= 2) {
    return {
      status: 'warning',
      message: `${Math.round(hoursRemaining * 10) / 10}h left`,
      color: '#fb923c',
      canTrade: true
    };
  }
  
  return {
    status: 'open',
    message: `${Math.round(hoursRemaining * 10) / 10}h left`,
    color: '#1de98b',
    canTrade: true
  };
}

export function formatPosition(position) {
  if (position === 0) return { text: 'FLAT', color: '#64748b', emoji: '➖' };
  if (position > 0) return { text: `LONG ${position.toFixed(1)}MW`, color: '#38c0fc', emoji: '📈' };
  return { text: `SHORT ${Math.abs(position).toFixed(1)}MW`, color: '#f0455a', emoji: '📉' };
}

// ─── PHASE MANAGEMENT ───

export const PHASES = {
  FORECAST: 'FORECAST',   // Day-ahead forecast generation
  DA: 'DA',               // Day-ahead curve submission
  IDA1: 'IDA1',           // Intraday auction round 1
  IDA2: 'IDA2',           // Intraday auction round 2
  ID: 'ID',               // Continuous intraday trading (order-book pay-as-bid)
  REALTIME: 'REALTIME',   // Real-time BM phase (SP-by-SP)
  BM_OPEN: 'BM_OPEN',     // BM accepting bids for current SP
  BM_CLOSE: 'BM_CLOSE',   // BM cleared + settled for current SP
  RESULTS: 'RESULTS',     // End-of-day results & scoring
};

export const PHASE_DISPLAY_NAMES = {
  [PHASES.FORECAST]: 'Forecast',
  [PHASES.DA]: 'Day-Ahead Auction',
  [PHASES.IDA1]: 'Intraday Auction 1',
  [PHASES.IDA2]: 'Intraday Auction 2',
  [PHASES.ID]: 'Continuous Intraday',
  [PHASES.REALTIME]: 'Real-Time Delivery',
  [PHASES.BM_OPEN]: 'BM Open',
  [PHASES.BM_CLOSE]: 'BM Closed',
  [PHASES.RESULTS]: 'Settlement & Results',
};

export function getPhaseFromTime(currentTimeHour) {
  if (currentTimeHour < 0.5) return PHASES.FORECAST;
  if (currentTimeHour < 1.5) return PHASES.DA;
  if (currentTimeHour < 2.5) return PHASES.IDA1;
  if (currentTimeHour < 3.5) return PHASES.IDA2;
  if (currentTimeHour < 23) return PHASES.ID;
  if (currentTimeHour < 24) return PHASES.REALTIME;
  return PHASES.RESULTS;
}

// ─── DA AUCTION TIMING ───

// Real GB market: 09:20 hourly, 15:30 half-hourly
// Game: Simplified to single 15:30 submission
export const DA_AUCTION_TIMES = [9.33, 15.5]; // 09:20 and 15:30 in decimal hours

export function getNextDAAuction(currentTimeHour) {
  for (const time of DA_AUCTION_TIMES) {
    if (currentTimeHour < time) {
      return time;
    }
  }
  // Next day
  return DA_AUCTION_TIMES[0] + 24;
}

export function isDAAuctionOpen(currentTimeHour) {
  // Open from 08:00 until 09:20 (hourly) and 14:00 until 15:30 (half-hourly)
  const isHourlyWindow = currentTimeHour >= 8 && currentTimeHour < 9.33;
  const isHalfHourlyWindow = currentTimeHour >= 14 && currentTimeHour < 15.5;
  return isHourlyWindow || isHalfHourlyWindow;
}

File: backend/engine/id_trading_engine.py
"""
GridForge ID Trading Engine — Continuous intraday order-book market.

Models GB EPEX SPOT / Nord Pool intraday:
  - Order-book, pay-as-bid: each trade executes at the aggressor's limit price
  - Per-SP gate closure: trading stops 1 hour before SP delivery
  - Multiple trading rounds between IDA2 and BM
  - Players post limit buy/sell orders for specific SPs
"""

from __future__ import annotations
import time
import random

from .constants import SP_DURATION_H, SPS_PER_DAY
from .utils import sp_time


# ─── GATE CLOSURE TIMING ───

def get_gate_closure_hour(sp: int) -> float:
    if 1 <= sp <= 20:
        return 23 + (sp - 1) * 0.5
    else:
        return 9 + (sp - 21) * 0.5


def get_bm_gate_closure_hour(sp: int) -> float:
    sp_start_hour = ((sp - 1) * 0.5) % 24
    bm_gate_hour = sp_start_hour - 1
    if bm_gate_hour < 0:
        bm_gate_hour += 24
    if sp <= 2:
        return 23 + (sp - 1) * 0.5
    return bm_gate_hour


def format_gate_closure_time(sp: int) -> str:
    hour = get_gate_closure_hour(sp)
    day = "TODAY" if sp <= 20 else "TOMORROW"
    h = int(hour) % 24
    m = int((hour % 1) * 60)
    return f"{h:02d}:{m:02d} {day}"


# ─── GATE STATUS ───

def is_id_gate_open(sp: int, current_time_hour: float) -> bool:
    gate_hour = get_gate_closure_hour(sp)
    if sp > 20:
        return current_time_hour >= 24 and current_time_hour < gate_hour + 24
    return current_time_hour < gate_hour


def get_time_to_gate_closure(sp: int, current_time_hour: float) -> float:
    gate_hour = get_gate_closure_hour(sp)
    if sp > 20:
        return (gate_hour + 24) - current_time_hour
    return gate_hour - current_time_hour


def get_open_sps(current_time_hour: float) -> list[int]:
    return [sp for sp in range(1, 49) if is_id_gate_open(sp, current_time_hour)]


def get_newly_closed_sps(current_time_hour: float, prev_time_hour: float) -> list[int]:
    return [
        sp for sp in range(1, 49)
        if is_id_gate_open(sp, prev_time_hour) and not is_id_gate_open(sp, current_time_hour)
    ]


# ─── ID TRADE STRUCTURE ───

def create_id_trade(sp: int, player_id: str, side: str, volume_mw: float, price: float, timestamp: float) -> dict:
    rand_suffix = "".join(random.choices("abcdefghijklmnopqrstuvwxyz0123456789", k=3))
    return {
        "id": f"id_{int(time.time() * 1000)}_{rand_suffix}",
        "sp": sp,
        "playerId": player_id,
        "side": side,
        "volumeMW": abs(volume_mw),
        "price": price,
        "timestamp": timestamp,
        "status": "pending",
    }


# ─── POSITION TRACKING ───

def calculate_positions(da_volumes: list[float], id_trades: list[dict], player_id: str) -> list[dict]:
    positions = []
    for sp in range(1, 49):
        da_vol = da_volumes[sp - 1] if sp - 1 < len(da_volumes) else 0
        sp_trades = [t for t in id_trades if t["sp"] == sp and t["playerId"] == player_id and t.get("status") == "confirmed"]

        id_buy_vol = sum(t["volumeMW"] for t in sp_trades if t["side"] == "buy")
        id_sell_vol = sum(t["volumeMW"] for t in sp_trades if t["side"] == "sell")

        net_position = da_vol + id_buy_vol - id_sell_vol

        positions.append({
            "sp": sp,
            "daVolume": da_vol,
            "idBuyVolume": id_buy_vol,
            "idSellVolume": id_sell_vol,
            "netPosition": net_position,
            "trades": sp_trades,
            "isOpen": True,
        })
    return positions


# ─── ID MARKET CLEARING ───

def match_id_orders(buy_orders: list[dict], sell_orders: list[dict]) -> list[dict]:
    """
    Order-book matching with pay-as-bid pricing.

    Sorts buys descending by price, sells ascending.  Walks the book:
    when best-buy >= best-sell, a trade occurs at the **earlier** order's
    price (i.e. the passive/resting side).  This is standard exchange
    convention — the aggressor (later order) gets the resting price.

    For the game we simplify: use the sell price for the trade
    (seller sets the floor, buyer pays that floor if willing).
    """
    matches = []
    sorted_buys = sorted(buy_orders, key=lambda b: b["price"], reverse=True)
    sorted_sells = sorted(sell_orders, key=lambda s: s["price"])

    si = 0
    for buy in sorted_buys:
        while si < len(sorted_sells):
            sell = sorted_sells[si]
            if buy["price"] < sell["price"]:
                break  # no more matchable sells
            if buy.get("volumeMW", buy.get("mw", 0)) <= 0:
                break
            if sell.get("volumeMW", sell.get("mw", 0)) <= 0:
                si += 1
                continue

            buy_vol = buy.get("volumeMW", buy.get("mw", 0))
            sell_vol = sell.get("volumeMW", sell.get("mw", 0))
            match_volume = min(buy_vol, sell_vol)

            # Pay-as-bid: trade at the passive (sell) side price
            match_price = sell["price"]

            matches.append({
                "sp": buy.get("sp", sell.get("sp")),
                "buyerId": buy.get("playerId", buy.get("id")),
                "sellerId": sell.get("playerId", sell.get("id")),
                "volumeMW": match_volume,
                "price": round(match_price, 2),
                "buyerPays": round(match_volume * match_price * SP_DURATION_H, 2),
                "sellerReceives": round(match_volume * match_price * SP_DURATION_H, 2),
            })

            # Reduce residual volume
            if "volumeMW" in buy:
                buy["volumeMW"] -= match_volume
            else:
                buy["mw"] = buy.get("mw", 0) - match_volume
            if "volumeMW" in sell:
                sell["volumeMW"] -= match_volume
            else:
                sell["mw"] = sell.get("mw", 0) - match_volume

            if sell.get("volumeMW", sell.get("mw", 0)) <= 0:
                si += 1

    return matches


# ─── CLEAR A FULL ID ROUND (all open SPs) ───

def clear_id_round(
    orders_by_player: dict,
    open_sps: list[int] | None = None,
) -> dict:
    """
    Run one round of continuous ID clearing across all open SPs.

    orders_by_player: { pid: [ { sp, side, mw, price }, ... ] }
    open_sps: list of SPs still open for trading (None = all 48)

    Returns:
      {
        "trades": [ { sp, buyerId, sellerId, volumeMW, price, ... } ],
        "tradesBySp": { sp: [...] },
        "totalVolume": float,
        "positionDeltas": { pid: { sp: delta_mw } },
        "cashDeltas": { pid: float },
      }
    """
    if open_sps is None:
        open_sps = list(range(1, SPS_PER_DAY + 1))
    open_set = set(open_sps)

    # Bucket orders by SP
    buys_by_sp: dict[int, list] = {}
    sells_by_sp: dict[int, list] = {}

    for pid, orders in orders_by_player.items():
        if not isinstance(orders, list):
            orders = [orders]
        for order in orders:
            sp = order.get("sp")
            if sp is None or sp not in open_set:
                continue
            entry = {
                **order,
                "playerId": pid,
                "volumeMW": float(order.get("mw", order.get("volumeMW", 0))),
            }
            side = order.get("side", "").lower()
            if side in ("buy", "bid"):
                buys_by_sp.setdefault(sp, []).append(entry)
            elif side in ("sell", "offer"):
                sells_by_sp.setdefault(sp, []).append(entry)

    all_trades = []
    trades_by_sp: dict[int, list] = {}
    pos_deltas: dict[str, dict[int, float]] = {}  # pid → { sp → MW delta }
    cash_deltas: dict[str, float] = {}  # pid → £ delta

    for sp in open_sps:
        sp_buys = buys_by_sp.get(sp, [])
        sp_sells = sells_by_sp.get(sp, [])
        if not sp_buys or not sp_sells:
            continue

        sp_trades = match_id_orders(sp_buys, sp_sells)
        if not sp_trades:
            continue

        all_trades.extend(sp_trades)
        trades_by_sp[sp] = sp_trades

        for t in sp_trades:
            vol = t["volumeMW"]
            payment = t.get("buyerPays", vol * t["price"] * SP_DURATION_H)

            # Buyer: increases position (buys power), pays cash
            buyer = t["buyerId"]
            pos_deltas.setdefault(buyer, {})[sp] = (
                pos_deltas.get(buyer, {}).get(sp, 0) - vol
            )
            cash_deltas[buyer] = cash_deltas.get(buyer, 0) - payment

            # Seller: decreases position (sells power), receives cash
            seller = t["sellerId"]
            pos_deltas.setdefault(seller, {})[sp] = (
                pos_deltas.get(seller, {}).get(sp, 0) + vol
            )
            cash_deltas[seller] = cash_deltas.get(seller, 0) + payment

    return {
        "trades": all_trades,
        "tradesBySp": trades_by_sp,
        "totalVolume": sum(t["volumeMW"] for t in all_trades),
        "positionDeltas": pos_deltas,
        "cashDeltas": cash_deltas,
    }


# ─── UI HELPERS ───

def get_gate_status_display(sp: int, current_time_hour: float) -> dict:
    is_open = is_id_gate_open(sp, current_time_hour)
    hours_remaining = get_time_to_gate_closure(sp, current_time_hour)

    if not is_open:
        return {"status": "closed", "message": "LOCKED - No more ID trades", "color": "#f0455a", "canTrade": False}

    if hours_remaining <= 0.5:
        return {"status": "urgent", "message": f"{round(hours_remaining * 60)}m left - URGENT", "color": "#f5b222", "canTrade": True}

    if hours_remaining <= 2:
        return {"status": "warning", "message": f"{round(hours_remaining * 10) / 10}h left", "color": "#fb923c", "canTrade": True}

    return {"status": "open", "message": f"{round(hours_remaining * 10) / 10}h left", "color": "#1de98b", "canTrade": True}


def format_position(position: float) -> dict:
    if position == 0:
        return {"text": "FLAT", "color": "#64748b", "emoji": "➖"}
    if position > 0:
        return {"text": f"LONG {position:.1f}MW", "color": "#38c0fc", "emoji": "📈"}
    return {"text": f"SHORT {abs(position):.1f}MW", "color": "#f0455a", "emoji": "📉"}


# ─── PHASE MANAGEMENT ───

PHASES = {
    "DA": "DA",
    "ID": "ID",
    "BM_GATE": "BM_GATE",
    "DELIVERY": "DELIVERY",
    "SETTLEMENT": "SETTLEMENT",
}

PHASE_DISPLAY_NAMES = {
    "DA": "Day-Ahead Auction",
    "ID": "Intraday Trading",
    "BM_GATE": "BM Gate Closed",
    "DELIVERY": "Live Delivery",
    "SETTLEMENT": "Settlement",
}


def get_phase_from_time(current_time_hour: float) -> str:
    if current_time_hour < 2:
        return PHASES["DA"]
    if current_time_hour < 23:
        return PHASES["ID"]
    return PHASES["ID"]


# ─── DA AUCTION TIMING ───

DA_AUCTION_TIMES = [9.33, 15.5]


def get_next_da_auction(current_time_hour: float) -> float:
    for t in DA_AUCTION_TIMES:
        if current_time_hour < t:
            return t
    return DA_AUCTION_TIMES[0] + 24


def is_da_auction_open(current_time_hour: float) -> bool:
    is_hourly = 8 <= current_time_hour < 9.33
    is_half_hourly = 14 <= current_time_hour < 15.5
    return is_hourly or is_half_hourly
